Add basic support for Zookeeper Serversets for SD

brian-brazil opened the pull request:
It can discover an entire tree of serversets, or just one.
This is a very naieve approach using polling,
this should be changed to use watches in the future.


For anyone reading this in future: This is not how you're
meant to use Zookeeper. Watches are far more efficient.

@fabxc

fabxc commented:
I understand serversets are a defined schemes for doing SD in ZK, which is great as it gives us something we can support.

Several concerns here:
- I'm not familiar with serversets. Does this cover the whole topic of serversets? Are they always JSON – if not which other encodings? If there's anything serverset SD could be extended with, will this break the current config?
- Google shows very few results for the whole serverset thing. It does not seem like something very common to me. I'd like to see more people expressing demand for it.
- `this should be changed to use watches in the future.` – by whom? Once this hits master it will always be at the bottom of someones backlog.
- The implementation does not implement target provider interface, it circumvents it.

Each discovery mechanism comes with an overhead (config, dependencies, code). We have to establish some requirements for native support. Those should be:

- Mechanism is a legitimate way of doing SD [✓]
- Sufficient demand [✗]
- Sound implementation [✗] that interacts with our target manager as intended [✗]

Confd + FileSD yields better results here.

brian-brazil commented:
> I'm not familiar with serversets. Does this cover the whole topic of serversets? Are they always JSON – if not which other encodings? If there's anything serverset SD could be extended with, will this break the current config?

They can also be Thrift, but it seems most people use JSON.

> Google shows very few results for the whole serverset thing. It does not seem like something very common to me. I'd like to see more people expressing demand for it.

It's common for those using the Twitter stack, such as Finagle and Aurora. I've already had a few people interested in it.

> The implementation does not implement target provider interface, it circumvents it.

Ah, I forgot to finish off that bit.

brian-brazil commented:
Okay, it took some work but this is using watches now.

brian-brazil commented:
That should be good now.

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+	return &labels, nil
+}
+
+type ZookeeperTreeCache struct {
```
This one and the one below can be unexported, right?

brian-brazil replied:
Yes, I'm planning on offering this upstream as an initial Go version of https://curator.apache.org/apidocs/org/apache/curator/framework/recipes/cache/TreeCache.html so I was setting visibility accordingly.

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+		conf:      conf,
+		conn:      conn,
+		updates:   updates,
+		sources:   make(map[string]struct{}),
```
We have the "no `make()` unless necessary" style guideline, so `map[string]struct{}{}`.

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+		return err
+	}
+
+	currentChildren := make(map[string]struct{})
```
`map[string]struct{}{}`

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+		if childNode == nil {
+			node.children[child] = &zookeeperTreeCacheNode{
+				events:   node.events,
+				children: make(map[string]*zookeeperTreeCacheNode),
```
dito

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+					if childNode == nil {
+						childNode = &zookeeperTreeCacheNode{
+							events:   tc.head.events,
+							children: make(map[string]*zookeeperTreeCacheNode),
```
dito

fabxc commented on retrieval/discovery/serverset.go:
```diff
@@ -0,0 +1,359 @@
+	Port int
+}
+
+type ZookeeperLogger struct {
```
Can be removed with the next vendor upgrade of our log package.

fabxc commented:
We manually re-insert LICENSE files and such into the vendoring directory – not entirely sure why the NOTICE file doesn't suffice legally either.

fabxc commented:
Wow, I didn't expect it to be this complicated since it boiled down to [this](https://github.com/fabxc/prom_sd_example/blob/master/etcd/main.go#L96-L202) for etcd.
So thanks a ton for the effort!

Aside from minor comments :+1: for the code – obviously haven't tested it myself.

beorn7 commented:
@fabxc The details about the license inclusion recursively depends on the license itself. In doubt, we must err on the side of including it.

brian-brazil commented:
I believe I've addressed all comments, and reworked the run/sources handling.

> Wow, I didn't expect it to be this complicated

Why do you think I started with a polling approach :)

fabxc commented:
Great :+1:
